File: Claude/Python/2025/Day11_Solution_2025.py
import time
import sys
from functools import lru_cache

start = time.perf_counter()

sys.setrecursionlimit(100000)

graph = {}
with open("Day11_Input.txt") as f:
    for line in f:
        line = line.strip()
        if not line:
            continue
        node, rest = line.split(': ', 1)
        targets = rest.split()
        graph[node] = targets

@lru_cache(maxsize=None)
def count_paths(node):
    if node == 'out':
        return 1
    return sum(count_paths(t) for t in graph.get(node, []))

print(count_paths("you"))

end = time.perf_counter()
print(f"Runtime: {end - start:.6f} seconds")

File: Claude/Rust/2025/Day11_Solution_2025.rs
use std::collections::HashMap;
use std::fs;

fn count_paths(node: &str, graph: &HashMap<String, Vec<String>>, memo: &mut HashMap<String, u128>) -> u128 {
    if node == "out" {
        return 1;
    }
    if let Some(&cached) = memo.get(node) {
        return cached;
    }
    let count = match graph.get(node) {
        Some(children) => children
            .iter()
            .map(|child| count_paths(child, graph, memo))
            .sum(),
        None => 0,
    };
    memo.insert(node.to_string(), count);
    count
}

fn main() {
    let content = fs::read_to_string("Day11_input.txt").expect("Could not read input file");

    let mut graph: HashMap<String, Vec<String>> = HashMap::new();

    for line in content.lines().filter(|l| !l.is_empty()) {
        if let Some((node, rest)) = line.split_once(": ") {
            let children: Vec<String> = rest.split_whitespace().map(str::to_string).collect();
            graph.insert(node.to_string(), children);
        }
    }

    let mut memo: HashMap<String, u128> = HashMap::new();
    let result = count_paths("you", &graph, &mut memo);

    println!("{}", result);
}
